### PR TITLE
Normalizing Flow Torch Model

### LINKED ISSUE #2863: TensorFlow -> PyTorch Model Conversion List
We've made a call to migrate our existing TensorFlow models in PyTorch for better longterm sustainability. Here is a list of TensorFlow models that have yet to be migrated

- [ ] `TensorflowMultitaskIRVClassifier`
- [ ] `RobustMultitaskClassifier`
- [ ] `RobustMultitaskRegressor`
- [ ] `ProgressiveMultitaskClassifier`
- [ ] `ProgressiveMultitaskRegressor`
- [ ] `WeaveModel`
- [ ] `DTNNModel`
- [ ] `DAGModel`
- [ ] `GraphConvModel`
- [x] `MPNNModel`
- [ ] `BasicMolGANModel`
- [ ] `ScScoreModel`
- [ ] `SeqToSeq`
- [ ] `GAN`, `WGAN`
- [ ] `CNN`
- [ ] `TextCNNModel`
- [ ] `AtomicConvModel`
- [ ] `Smiles2Vec`
- [ ] `AtomicConvModel`
- [ ] `ChemCeption`
- [ ] `NormalizingFlowModel`

When doing a conversion, I recommend first implementing the new model in `TorchModel` and writing tests to make sure that the qualitative and quantitative behavior of the new model matches the TensorFlow base implementation. We will migrate models to replace the old implementation over a release or two

### PR DESCRIPTION
**Pull request by JoseAntonioSiguenza**
# Pull Request

## Description

Fix #2863 

The creation of this model would solve one of the listed models to be ported at #2863. Also, to complete this model, a class that considers Deepchem's API must be added.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 1]
```diff
@@ -1,2 +1,65 @@
...
+import torch
+from torch import nn
+from typing import Sequence, Tuple
+
+
+class NormalizingFlow(nn.Module):
+  """Normalizing flows are widley used to perform generative models.
+  Normalizing flow models gives advantages over variational autoencoders
+  (VAE) because of ease in sampling by applying invertible transformations
+  (Frey, Gadepally, & Ramsundar, 2022)."""
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:12]: Can you add a usage example here?

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 2]
```diff
@@ -1,2 +1,65 @@
 """Normalizing flows for transforming probability distributions using PyTorch.
 """
+import torch
+from torch import nn
+from typing import Sequence, Tuple
+
+
+class NormalizingFlow(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:8]: Can you add some additional unit tests?

File: deepchem/models/tests/test_normalizing_flow_pytorch.py [hunk 3]
```diff
@@ -44,3 +45,48 @@ def test_Affine():
 
   assert np.array_equal(log_det_jacobian, zeros)
   assert np.array_equal(inverse_log_det_jacobian, zeros)
+
+
+def test_Normalizing_Flow_Pytorch():
```
Comment (rbharath) [inline deepchem/models/tests/test_normalizing_flow_pytorch.py:50]: Should `test_normalizing_flow_pytorch`

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 4]
```diff
@@ -1,2 +1,89 @@
...
+      log_prob += inverse_log_det_jacobian
+
+    return log_prob
+
+  def sample(self, n_samples: int) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Performs a sampling from the transformed distribution.
+    Besides the outputs (sampling), this method returns the logarithm of
+    probability to obtain the outputs at the base distribution.
+
+    n_samples shape: (samples)
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:78]: Needs to be formatted in numpydoc style https://numpydoc.readthedocs.io/en/latest/format.html

File: deepchem/models/tests/test_normalizing_flow_pytorch.py [hunk 5]
```diff
@@ -44,3 +45,48 @@ def test_Affine():
...
+
+    log_prob: Method used to learn parameter (optimizing loop)
+      input shape: (samples)
+      output shape: (samples)
+
+  """
+
+  dim = 2
+  samples = 96
+  base_distribution = MultivariateNormal(torch.zeros(dim), torch.eye(dim))
```
Comment (rbharath) [inline deepchem/models/tests/test_normalizing_flow_pytorch.py:70]: `MultivariateNormal` is not imported at the top

**Comment by rbharath** [comment]
@ncfrey Would you be able to take a look at the proposed normalizing flow API in this PR to see if it makes sense?

**Review by ncfrey — COMMENTED** [review]
Nicely done! A few general comments:
1. You may want a base class for transforms that has forward and inverse methods that all transforms will inherit from.
2. If you haven't looked at the [nflows](https://github.com/bayesiains/nflows) library, I highly recommend it. They have a lot of transforms implemented and a solid API design.
3. I think basic normalizing flows struggle pretty mightily on distribution learning for molecules, whether the representation is a string or a graph. One solution could be to use much more flexible base distributions than a Multivariate Normal, e.g. use a VAE prior. Something to keep in mind to make these models more functional. There are also many applications beyond generating molecules that are potentially more interesting (integrations with FEP and MD).

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 6]
```diff
@@ -1,2 +1,89 @@
...
 """
+import torch
+from torch import nn
+from typing import Sequence, Tuple
+
+
+class NormalizingFlow(nn.Module):
+  """Normalizing flows are widley used to perform generative models.
+  This algorithm gives advantages over variational autoencoders (VAE) because
+  of ease in sampling by applying invertible transformations
```
[the remark this answers is not in the document]
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/normalizing_flows_pytorch.py:11]: Suree!! Definitely considering FastFlows paper 🙌🏼

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 7]
```diff
@@ -1,2 +1,89 @@
...
+  >>> tensor = distribution.sample(torch.Size((samples, dim)))
+  >>> len(model.log_prob(tensor))
+  96
+  >>> # evaluates the the sampling method and its log_prob
+  >>> len(model.sample(samples))
+  2
+
+  """
+
+  def __init__(self, transform: Sequence, base_distribution, dim: int) -> None:
```
[the remark this answers is not in the document]
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/normalizing_flows_pytorch.py:38]: Yes, so I have some time on deciding which type hint class to write here. I looked that some CI fails raised If used `torch.Tensor`. This is mainly because `base_distribution` is a `torch.distributions` object. I will write `torch.distributions`, does this make sense 🤔?

**Comment by JoseAntonioSiguenza** [comment]
> Nicely done! A few general comments:
> 
> 1. You may want a base class for transforms that has forward and inverse methods that all transforms will inherit from.
> 2. If you haven't looked at the [nflows](https://github.com/bayesiains/nflows) library, I highly recommend it. They have a lot of transforms implemented and a solid API design.
> 3. I think basic normalizing flows struggle pretty mightily on distribution learning for molecules, whether the representation is a string or a graph. One solution could be to use much more flexible base distributions than a Multivariate Normal, e.g. use a VAE prior. Something to keep in mind to make these models more functional. There are also many applications beyond generating molecules that are potentially more interesting (integrations with FEP and MD).

Thanks, @ncfrey!! I really appreciate feedback 🙌🏼. At the beginning of this PR, I considered having a base `Transform` object as you suggest, but I thought to add those methods for each layer. Nevertheless, with this being said I will include a base class at this PR. About the third comment, I fully agree with you. Also, I've looked that a mix of transformation layers might improve results. Finally, normalizing flows are really cool and I'm going to read more about its applicability with FEP and MD.

File: deepchem/models/torch_models/normalizing_flows_pytorch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/normalizing_flows_pytorch.py:38]: Awesome!! I will Include this type hint as well as the `Transforms` class and FastFlow reference on my next PR 🙌🏼

**Comment by JoseAntonioSiguenza** [comment]
Hi @arunppsg 👋🏽! Could you please confirm if these CI errors are not related to the modified code and also if this PR can get merged in without conflict 🙌🏼

**Comment by arunppsg** [comment]
The CI errors are not related to this PR. Is this good to merge?

**Comment by rbharath** [comment]
Good to go on my side. @JoseAntonioSiguenza Good to go?
